### PR TITLE
dp and yang push demo/example

### PR DESCRIPTION
**Pull request by micnovak**
Demo/example showing YANG Push driven by separate thread in external DP (partly in C++).

### CODE REVIEW

File: dp_and_yang_push/Makefile [hunk 1]
```diff
@@ -0,0 +1,102 @@
+######################################################################
+# ConfD OpenVPN  - demo
+# (C) 2006-2018 Tail-f Systems
```
Comment (jomiklos) [inline dp_and_yang_push/Makefile]: deprecated...

File: dp_and_yang_push/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Data provider demo with YANG Push
```
Comment (jomiklos) [inline dp_and_yang_push/README.md]: "_and_" feels redundant/strange in the directory name...

File: dp_and_yang_push/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+functionality. The operational data changes are driven by example randomly in a separate thread. 
+
+The example is partly written in C++ (mainly to use `thread`, `string`, `vector` and `map`). 
+It also demonstrates how to use C++ with C ConfD API and use it in a ConfD Application.
+
+In addition, simple performance test is provided to test data provider speed for various amount of data.
+
+The data provider itself implements `get_elem`, `get_next`, `get_object` and `get_next_object` callbacks. For YANG-Push support example implements `subscribe_on_change` and `unsubscribe_on_change` callbacks.  
+ 
+The example reuses some code form ConfD example found in `examples.confd/netconf_yang_push`.
```
Comment (jomiklos) [inline dp_and_yang_push/README.md]: "form" -> "from"

File: dp_and_yang_push/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+* to periodically fetch operational data, use `make periodic-subscription`
+
+## Stop
+
+* use `make stop`
+
+## Performance
+
+For simple DP performance measurement run `./oper-get-provider-test.sh` 
+(this will  build, start application, run measurement, stop and clean application).
```
Comment (jomiklos) [inline dp_and_yang_push/README.md]: extra space: "will  build"

File: dp_and_yang_push/oper-get-provider-test.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+NUMLE_START=$(echo "1" | bc)
```
Comment (jomiklos) [inline dp_and_yang_push/oper-get-provider-test.sh]: what is "NUMLE"?

File: dp_and_yang_push/oper-get-provider-test.sh [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        delta_time "get 1 leaf from entry with ID ${VAL} out of $NUMLE"
+
+        echo "---netconf-console <get> - 1 list entry with ID ${VAL}"
+        START=$(date +%s.%N)
+        netconf-console --get -x /route-status/route[id=\"${VAL}rt\"] > dp_route-data-${NUMLE}-list-entry-with-id-${VAL}.xml
+        delta_time "get 1 list entry with ID ${VAL} out of $NUMLE"
+
+        echo "---netconf-console <get> - all records"
+        START=$(date +%s.%N)
+        netconf-console --get -x route-status > dp_route-data-${NUMLE}.xml
```
Comment (jomiklos) [inline dp_and_yang_push/oper-get-provider-test.sh]: mixing dashes & underscores in filename is ugly practice :(

File: dp_and_yang_push/route-status.yang [hunk 7]
```diff
@@ -0,0 +1,33 @@
+
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.yang]: empty leading line...

File: dp_and_yang_push/route-status.cpp [hunk 8]
```diff
@@ -0,0 +1,575 @@
+/*********************************************************************
+ * ConfD Stats intro example
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.cpp]: dead comment...

File: dp_and_yang_push/route-status.cpp [hunk 9]
```diff
@@ -0,0 +1,575 @@
...
+    trans.finish = s_finish;
+
+    memset(&data, 0, sizeof(struct confd_data_cbs));
+    data.get_elem = get_elem;
+    data.get_next = get_next;
+    data.get_object = get_object;
+    data.get_next_object = get_next_object;
+    strcpy(data.callpoint, rs__callpointid_routestat);
+    const char *daemon_name = "router_daemon";
+    /* initialize confd library */
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.cpp]: redundant comment

File: dp_and_yang_push/route-status.cpp [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+
+    if (confd_register_done(dctx) != CONFD_OK)
+        confd_fatal("Failed to complete registration \n");
+    INFO_EXIT("");
+}
+
+void generate_changes()
+{
+    INFO_ENTER("");
+    TRACE("Num of routes=%i", routes.size());
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.cpp]: missing spaces around =

File: dp_and_yang_push/route-status.cpp [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+
+    end:
+    TRACE_EXIT("ret %i", ret);
+    return ret;
+}
+
+static int get_next_object(struct confd_trans_ctx *tctx,
+        confd_hkeypath_t *keypath, long next)
+{
+    TRACE_ENTER("next %i routes.size() %u", next, routes.size());
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.cpp]: cryptic "sentence", add e.g. ; separator?

File: dp_and_yang_push/Makefile [hunk 12]
```diff
@@ -0,0 +1,102 @@
+######################################################################
```
Comment (jomiklos) [inline dp_and_yang_push/Makefile]: whole "ConfD-Developer" root repository has only dashes in directory names, this is the first and only one introducing underscores... maybe reconsider naming for this example root dir?

File: dp_and_yang_push/route-status.cpp [hunk 13]
```diff
@@ -0,0 +1,575 @@
...
+/********************************************************************/
+/* Helper functions */
+
+void print_val(const char *str, const confd_value_t *val)
+{
+#ifdef T_LOG_TRACE
+    if (val == NULL) {
+        TRACE("%s == NULL", str);
+    } else {
+        int buff_len = 256;
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.cpp]: const?

File: dp_and_yang_push/route-status.cpp [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+int main(int argc, char *argv[])
+{
+    INFO_ENTER("argc %i", argc);
+    int ret = CONFD_OK;
+    bool arg_ok = true;
+
+    if (argc != 2) {
+        arg_ok = false;
+    }
+    try {
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.cpp]: bit broken - arg_ok could be already false at this point

File: dp_and_yang_push/route-status.cpp [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+    datetime->timezone = 0;
+    datetime->timezone_minutes = 0;
+    datetime->hour = tm.tm_hour;
+    datetime->min = tm.tm_min;
+    TRACE_EXIT("");
+}
+
+static void cb_push_on_change(route_type *rt)
+{
+    INFO_ENTER("");
```
Comment (jomiklos) [inline dp_and_yang_push/route-status.cpp]: mismatched in/out (INFO/TRACE)?

File: dp_and_yang_push/Makefile (hunk 12 above)
Comment (micnovak) [reply dp_and_yang_push/Makefile]: I like underscores more, but you are right, dashes are preffered these days.

File: dp_and_yang_push/README.md (hunk 2 above)
Comment (micnovak) [reply dp_and_yang_push/README.md]: will rename

File: dp_and_yang_push/oper-get-provider-test.sh (hunk 5 above)
Comment (micnovak) [reply dp_and_yang_push/oper-get-provider-test.sh]: will change to NUM

File: dp_and_yang_push/oper-get-provider-test.sh (hunk 6 above)
Comment (micnovak) [reply dp_and_yang_push/oper-get-provider-test.sh]: will change to dp-route

File: dp_and_yang_push/route-status.yang (hunk 7 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.yang]: removed

File: dp_and_yang_push/route-status.cpp (hunk 8 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.cpp]: removed

File: dp_and_yang_push/route-status.cpp (hunk 9 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.cpp]: removed

File: dp_and_yang_push/route-status.cpp (hunk 10 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.cpp]: I'm used to this for logging (space may be part of output)

File: dp_and_yang_push/route-status.cpp (hunk 11 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.cpp]: added `=` (not sentence)

File: dp_and_yang_push/route-status.cpp (hunk 13 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.cpp]: added `const`

File: dp_and_yang_push/route-status.cpp (hunk 14 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.cpp]: will fix

File: dp_and_yang_push/route-status.cpp (hunk 15 above)
Comment (micnovak) [reply dp_and_yang_push/route-status.cpp]: chaged to INFO

**Review by cohult — COMMENTED** [review]
Seems like there are two folders with the same demo:
dp-yang-push
dp_and_yang_push
One of them should be deleted?

**Pull request merged**